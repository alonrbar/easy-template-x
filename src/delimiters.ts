
export class Delimiters {

    public tagStart = "{";
    public tagEnd = "}";
    public containerTagOpen = "#";
    public containerTagClose = "/";
    public tagOptionsStart = "[";
    public tagOptionsEnd = "]";

    constructor(initial?: Partial<Delimiters>) {
        Object.assign(this, initial);

        this.encodeAndValidate();

        if (this.containerTagOpen === this.containerTagClose)
            throw new Error(`containerTagOpen can not be equal to containerTagClose`);
    }

    private encodeAndValidate() {
        const keys: (keyof Delimiters)[] = ['tagStart', 'tagEnd', 'containerTagOpen', 'containerTagClose'];
        for (const key of keys) {

            const value = this[key];
            if (!value)
                throw new Error(`${key} can not be empty.`);

            if (value !== value.trim())
                throw new Error(`${key} can not contain leading or trailing whitespace.`);
        }
    }
}
