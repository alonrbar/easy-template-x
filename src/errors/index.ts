export * from './internalArgumentMissingError';
export * from './internalError';
export * from './malformedFileError';
export * from './maxXmlDepthError';
export * from './missingCloseDelimiterError';
export * from './missingStartDelimiterError';
export * from './tagOptionsParseError';
export * from './templateDataError';
export * from './templateSyntaxError';
export * from './unclosedTagError';
export * from './unidentifiedFileTypeError';
export * from './unknownContentTypeError';
export * from './unopenedTagError';
export * from './unsupportedFileTypeError';
