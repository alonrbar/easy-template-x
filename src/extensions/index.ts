export * from "./IExtensionUtilities";
export * from "./ITemplateExtension";
export * from "./ITemplatePlugin";
export * from "./contentControlExtension";
export * from "./defaultExtensions";
export * from "./extensionUtilities";
export * from "./templateExtension";
