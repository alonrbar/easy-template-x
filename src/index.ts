export * from './compilation';
export * from './errors';
export * from './office';
export * from './plugins';
export * from './utils';
export * from './xml';
export * from './zip';
export * from './delimiters';
export * from './mimeType';
export * from './templateData';
export * from './templateHandler';
export * from './templateHandlerOptions';
