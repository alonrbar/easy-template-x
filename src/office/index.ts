export * from './contentPartType';
export * from './docx';
export * from './docxParser';
export * from './relationship';
export * from './xmlPart';

