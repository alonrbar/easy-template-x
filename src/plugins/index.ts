export * from './image';
export * from './link';
export * from './loop';
export * from './rawXml';
export * from './text';
export * from "./chart";
export * from './defaultPlugins';
export * from './pluginContent';
export * from './templatePlugin';
