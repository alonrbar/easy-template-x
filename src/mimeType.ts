import { UnsupportedFileTypeError } from "./errors";
import { RelType } from "./office/relationship";

export const MimeType = Object.freeze({
    Png: 'image/png',
    Jpeg: 'image/jpeg',
    Gif: 'image/gif',
    Bmp: 'image/bmp',
    Svg: 'image/svg+xml'
} as const);

export type MimeType = typeof MimeType[keyof typeof MimeType];

export class MimeTypeHelper {

    public static getDefaultExtension(mime: MimeType): string {
        switch (mime) {
            case MimeType.Png:
                return 'png';
            case MimeType.Jpeg:
                return 'jpg';
            case MimeType.Gif:
                return 'gif';
            case MimeType.Bmp:
                return 'bmp';
            case MimeType.Svg:
                return 'svg';
            default:
                throw new UnsupportedFileTypeError(mime);
        }
    }

    public static getOfficeRelType(mime: MimeType): string {
        switch (mime) {
            case MimeType.Png:
            case MimeType.Jpeg:
            case MimeType.Gif:
            case MimeType.Bmp:
            case MimeType.Svg:
                return RelType.Image;
            default:
                throw new UnsupportedFileTypeError(mime);
        }
    }
}
