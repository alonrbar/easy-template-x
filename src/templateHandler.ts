import { DelimiterSearcher, ScopeData, Tag, TagParser, TemplateCompiler, TemplateContext } from './compilation';
import { Delimiters } from './delimiters';
import { MalformedFileError } from './errors';
import { Docx, DocxParser } from './office';
import { TemplateHandlerOptions } from './templateHandlerOptions';
import { Binary, EASY_VERSION, Constructor } from './utils';
import { XmlNode, XmlParser } from './xml';
import { Zip } from './zip';

export class TemplateHandler {

    /**
     * Version number of the `easy-template-x` library.
     */
    public readonly version = (typeof EASY_VERSION !== 'undefined' ? EASY_VERSION : 'null');

    private readonly xmlParser = new XmlParser();
    private readonly docxParser: DocxParser;
    private readonly compiler: TemplateCompiler;

    private readonly options: TemplateHandlerOptions;

    constructor(options?: TemplateHandlerOptions) {
        this.options = new TemplateHandlerOptions(options);

        //
        // this is the library's composition root
        //

        this.docxParser = new DocxParser(this.xmlParser);

        const delimiterSearcher = new DelimiterSearcher(this.docxParser);
        delimiterSearcher.startDelimiter = this.options.delimiters.tagStart;
        delimiterSearcher.endDelimiter = this.options.delimiters.tagEnd;
        delimiterSearcher.maxXmlDepth = this.options.maxXmlDepth;

        const tagParser = new TagParser(this.docxParser, this.options.delimiters as Delimiters);

        this.compiler = new TemplateCompiler(
            delimiterSearcher,
            tagParser,
            this.options.plugins,
            this.options.defaultContentType,
            this.options.containerContentType
        );

        this.options.plugins.forEach(plugin => {
            plugin.setUtilities({
                xmlParser: this.xmlParser,
                docxParser: this.docxParser,
                compiler: this.compiler
            });
        });
    }

    public async process<T extends Binary>(templateFile: T, data: any): Promise<T> {

        // load the docx file
        const docx = await this.loadDocx(templateFile);
        const document = await docx.getDocument();

        // process content (do replacements)        
        const scopeData = new ScopeData(data);
        const context: TemplateContext = {
            docx
        };
        await this.compiler.compile(document, scopeData, context);

        // export the result
        return docx.export(templateFile.constructor as Constructor<T>);
    }

    public async parseTags(templateFile: Binary): Promise<Tag[]> {
        const docx = await this.loadDocx(templateFile);
        const document = await docx.getDocument();
        return this.compiler.parseTags(document);
    }

    /**
     * Get the text content of the main document file.
     */
    public async getText(docxFile: Binary): Promise<string> {
        const docx = await this.loadDocx(docxFile);
        const text = await docx.getDocumentText();
        return text;
    }

    /**
     * Get the xml tree of the main document file.
     */
    public async getXml(docxFile: Binary): Promise<XmlNode> {
        const docx = await this.loadDocx(docxFile);
        const document = await docx.getDocument();
        return document;
    }

    //
    // private methods
    //

    private async loadDocx(file: Binary): Promise<Docx> {

        // load the zip file
        let zip: Zip;
        try {
            zip = await Zip.load(file);
        } catch {
            throw new MalformedFileError('docx');
        }

        // load the docx file
        const docx = this.docxParser.load(zip);
        return docx;
    }
}