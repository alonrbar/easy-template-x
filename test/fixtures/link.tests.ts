import { TemplateHandler } from "src/templateHandler";
import { describe, expect, it } from "vitest";
import { readFixture } from "./fixtureUtils";

describe('link fixture', () => {

    it("inserts hyper link", async () => {

        const template = readFixture("link.docx");

        const data = {
            link: {
                _type: 'link',
                text: "It's easy...",
                target: 'https://github.com/alonrbar/easy-template-x',
                tooltip: 'Click to open link'
            }
        };

        const handler = new TemplateHandler();
        const doc = await handler.process(template, data);

        const docXml = await handler.getXml(doc);
        expect(docXml).toMatchSnapshot();

        // writeTempFile('link - output.docx', doc);
    });
});
