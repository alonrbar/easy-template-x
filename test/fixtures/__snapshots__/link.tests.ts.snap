// Jest Snapshot v1, https://jestjs.io/docs/snapshot-testing

exports[`link fixture inserts hyper link 1`] = `<w:document xmlns:wpc="http://schemas.microsoft.com/office/word/2010/wordprocessingCanvas" xmlns:cx="http://schemas.microsoft.com/office/drawing/2014/chartex" xmlns:cx1="http://schemas.microsoft.com/office/drawing/2015/9/8/chartex" xmlns:cx2="http://schemas.microsoft.com/office/drawing/2015/10/21/chartex" xmlns:cx3="http://schemas.microsoft.com/office/drawing/2016/5/9/chartex" xmlns:cx4="http://schemas.microsoft.com/office/drawing/2016/5/10/chartex" xmlns:cx5="http://schemas.microsoft.com/office/drawing/2016/5/11/chartex" xmlns:cx6="http://schemas.microsoft.com/office/drawing/2016/5/12/chartex" xmlns:cx7="http://schemas.microsoft.com/office/drawing/2016/5/13/chartex" xmlns:cx8="http://schemas.microsoft.com/office/drawing/2016/5/14/chartex" xmlns:mc="http://schemas.openxmlformats.org/markup-compatibility/2006" xmlns:aink="http://schemas.microsoft.com/office/drawing/2016/ink" xmlns:am3d="http://schemas.microsoft.com/office/drawing/2017/model3d" xmlns:o="urn:schemas-microsoft-com:office:office" xmlns:r="http://schemas.openxmlformats.org/officeDocument/2006/relationships" xmlns:m="http://schemas.openxmlformats.org/officeDocument/2006/math" xmlns:v="urn:schemas-microsoft-com:vml" xmlns:wp14="http://schemas.microsoft.com/office/word/2010/wordprocessingDrawing" xmlns:wp="http://schemas.openxmlformats.org/drawingml/2006/wordprocessingDrawing" xmlns:w10="urn:schemas-microsoft-com:office:word" xmlns:w="http://schemas.openxmlformats.org/wordprocessingml/2006/main" xmlns:w14="http://schemas.microsoft.com/office/word/2010/wordml" xmlns:w15="http://schemas.microsoft.com/office/word/2012/wordml" xmlns:w16cid="http://schemas.microsoft.com/office/word/2016/wordml/cid" xmlns:w16se="http://schemas.microsoft.com/office/word/2015/wordml/symex" xmlns:wpg="http://schemas.microsoft.com/office/word/2010/wordprocessingGroup" xmlns:wpi="http://schemas.microsoft.com/office/word/2010/wordprocessingInk" xmlns:wne="http://schemas.microsoft.com/office/word/2006/wordml" xmlns:wps="http://schemas.microsoft.com/office/word/2010/wordprocessingShape" mc:Ignorable="w14 w15 w16se w16cid wp14"><w:body><w:p w14:paraId="604CB376" w14:textId="506341EE" w:rsidR="007E51D4" w:rsidRDefault="00B94C74" w:rsidP="00E14424"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t xml:space="preserve">Hi </w:t></w:r><w:hyperlink r:id="rId7" w:tooltip="Click to open link" w:history="1"><w:r><w:rPr><w:rStyle w:val="IntenseQuoteChar"/></w:rPr><w:rPr><w:rStyle w:val="Hyperlink"/></w:rPr><w:t>It&apos;s easy...</w:t></w:r></w:hyperlink><w:r w:rsidR="002632BB"><w:rPr><w:rStyle w:val="IntenseQuoteChar"/></w:rPr></w:r><w:r w:rsidR="007E51D4" w:rsidRPr="00825C5E"><w:rPr><w:rStyle w:val="IntenseQuoteChar"/></w:rPr></w:r><w:r><w:t xml:space="preserve"> hello</w:t></w:r></w:p><w:p w14:paraId="38AD1C23" w14:textId="641D2F59" w:rsidR="00B94C74" w:rsidRDefault="00B94C74" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr></w:p><w:tbl><w:tblPr><w:tblStyle w:val="TableGrid"/><w:tblW w:w="0" w:type="auto"/><w:tblLook w:val="04A0" w:firstRow="1" w:lastRow="0" w:firstColumn="1" w:lastColumn="0" w:noHBand="0" w:noVBand="1"/></w:tblPr><w:tblGrid><w:gridCol w:w="3005"/><w:gridCol w:w="3005"/><w:gridCol w:w="3006"/></w:tblGrid><w:tr w:rsidR="00B94C74" w14:paraId="7D668F01" w14:textId="77777777" w:rsidTr="00B94C74"><w:tc><w:tcPr><w:tcW w:w="3005" w:type="dxa"/></w:tcPr><w:p w14:paraId="6EFDA992" w14:textId="29EAE600" w:rsidR="00B94C74" w:rsidRDefault="00B94C74" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t>Hello</w:t></w:r></w:p></w:tc><w:tc><w:tcPr><w:tcW w:w="3005" w:type="dxa"/></w:tcPr><w:p w14:paraId="145F34B6" w14:textId="613F1DF3" w:rsidR="00B94C74" w:rsidRDefault="00B94C74" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr></w:p></w:tc><w:tc><w:tcPr><w:tcW w:w="3006" w:type="dxa"/></w:tcPr><w:p w14:paraId="2375E889" w14:textId="3981F582" w:rsidR="00B94C74" w:rsidRDefault="00825C5E" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t>123</w:t></w:r></w:p></w:tc></w:tr><w:tr w:rsidR="00B94C74" w14:paraId="2E1E4CF5" w14:textId="77777777" w:rsidTr="00B94C74"><w:tc><w:tcPr><w:tcW w:w="3005" w:type="dxa"/></w:tcPr><w:p w14:paraId="62892541" w14:textId="5A9A0164" w:rsidR="00B94C74" w:rsidRDefault="00DF322B" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t xml:space="preserve">Hi </w:t></w:r><w:hyperlink r:id="rId7" w:tooltip="Click to open link" w:history="1"><w:r><w:rPr><w:rStyle w:val="Hyperlink"/></w:rPr><w:t>It&apos;s easy...</w:t></w:r></w:hyperlink><w:r><w:t xml:space="preserve"> some</w:t></w:r></w:p></w:tc><w:tc><w:tcPr><w:tcW w:w="3005" w:type="dxa"/></w:tcPr><w:p w14:paraId="42670D3F" w14:textId="6A62D5A1" w:rsidR="00B94C74" w:rsidRDefault="00DF322B" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t xml:space="preserve">Some </w:t></w:r><w:hyperlink r:id="rId7" w:tooltip="Click to open link" w:history="1"><w:r><w:rPr><w:color w:val="FF0000"/></w:rPr><w:rPr><w:rStyle w:val="Hyperlink"/></w:rPr><w:t>It&apos;s easy...</w:t></w:r></w:hyperlink><w:r w:rsidRPr="00622901"><w:t xml:space="preserve"> </w:t></w:r><w:r><w:t>hi</w:t></w:r></w:p></w:tc><w:tc><w:tcPr><w:tcW w:w="3006" w:type="dxa"/></w:tcPr><w:p w14:paraId="0674138A" w14:textId="6CC118B0" w:rsidR="00B94C74" w:rsidRDefault="00B94C74" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr></w:p></w:tc></w:tr></w:tbl><w:p w14:paraId="4EAE51DB" w14:textId="77777777" w:rsidR="00B94C74" w:rsidRDefault="00B94C74" w:rsidP="00B94C74"><w:pPr><w:bidi w:val="0"/></w:pPr></w:p><w:p w14:paraId="71C40308" w14:textId="398848F5" w:rsidR="00A56ECC" w:rsidRDefault="00E6474A" w:rsidP="00A56ECC"><w:pPr><w:bidi w:val="0"/></w:pPr><w:r><w:t xml:space="preserve">Hi </w:t></w:r><w:hyperlink r:id="rId7" w:tooltip="Click to open link" w:history="1"><w:r><w:rPr><w:rStyle w:val="Hyperlink"/></w:rPr><w:t>It&apos;s easy...</w:t></w:r></w:hyperlink><w:r><w:t xml:space="preserve"> hello</w:t></w:r></w:p><w:p w14:paraId="5019EF3B" w14:textId="77777777" w:rsidR="00A56ECC" w:rsidRDefault="00A56ECC" w:rsidP="00A56ECC"><w:pPr><w:bidi w:val="0"/></w:pPr><w:bookmarkStart w:id="0" w:name="_GoBack"/><w:bookmarkEnd w:id="0"/></w:p><w:sectPr w:rsidR="00A56ECC" w:rsidSect="00C4055C"><w:pgSz w:w="11906" w:h="16838"/><w:pgMar w:top="1440" w:right="1440" w:bottom="1440" w:left="1440" w:header="708" w:footer="708" w:gutter="0"/><w:cols w:space="708"/><w:bidi/><w:rtlGutter/><w:docGrid w:linePitch="360"/></w:sectPr></w:body></w:document>`;
